doc/adr/0101-separating-concerns-between-persisters-actions.md

```diff
@@ -0,0 +1,133 @@
+* Persisters  
+  Store and load workflow state; in case of LedgerSMB, the state is stored in
+  a PostgreSQL database in a.o. the `workflow` and `workflow_context` tables
+* Observers  
+  Define Perl code to be triggered on events in a workflow
+
+#### Execution flow of `Workflow`
+
+1. Start database transaction
+2. Load workflow state from database
+3. Conditions are checked to select available actions from the list of all
+   actions defined in the current state
+4. An action is picked to be triggered
+5. Validators are run to check input validity for the selected action
+6. The selected action is run
+7. The workflow state is serialized to the database
+8. The database transaction is committed
+
+At various points in this flow observers are sent events from processing the
+action and optional state changes.
+
+### The problem with an active storage layer
+
+The `Workflow` state management library divides the concerns "performing actions"
+and "storing resulting workflow state" between the `Workflow::Action` and the
+`Workflow::Persister` respectively. Is the context of a passive storage backend
+(such as a filesystem), the consequences are straight forward: the persister
+serializes the context and stores it on disk. The rest happens in actions.
+
+In LedgerSMB, the storage layer isn't just a passive component: it's a
+PostgreSQL database server with stored procedures as active components. In case
+of a passive storage, access to it is restricted to the persister. In case of
+the active storage in LedgerSMB, this demarcation isn't as straight forward.
+
+Consider period closing, for example: there are two actions (`close` and
+`reopen`) which move the state of the workflow between states `CLOSED` and
+`OPEN`. The persister stores the workflow state, which is the end date of the
+closed period. Each workflow in `CLOSED` state should have an "accounting
+snapshot" stored. No workflow in `OPEN` state can have an accounting snapshot:
+the consequence of the `close` action hence should be the creation of a
+snapshot. If the `close` action itself is to ensure the creation of the
+snapshot, it needs access to the storage layer.
+
+There are two options to deal with the active side of the storage layer:
+
+1. Worklfow actions operate on the active side of the storage layer  
+   i.e. workflow actions trigger stored procedures
+2. Workflow persisters communicate with actions to make the persister
+   perform the necessary actions  
+   i.e. actions operate solely on workflow state, unable to access the
+   persistence layer
+
+The downside of (1) is that in its design, Workflow didn't account for
+actions which need to access the database (storage) layer, requiring
+customization of the workflow to support the pattern.
+
+The downside of (2) is that it results in tight coupling between actions
+and persisters: any new actions which require invocation of stored
+procedures will need changes in the persister as well.
```

Comment:
I'm trying to argue for a loose coupling here -- perhaps allow actions/conditions to get "auto-discovered" or registered in some dependency injection system or wired up through some external module, so the "persister" doesn't need to be modified at all to correctly support any number of actions

Reply:
The workflow is fully dependency-injected, or at least: actions are linked to code dynamically through configuration. The only reason a persister and an action can be mutually dependent is when it's the persister's role to do *all* database communication, where the action "only" adds flagging to the workflow state so the persister performs the expected database actions.

The route this PR is now taking is to have loose coupling: actions have access to the active layer of the database (stored procedures) while the persister has access to the passive layer of the database (tables).